### PR TITLE
James 1720 Handle ZonedDateTime correctly with Cassandra

### PR DESCRIPTION
**Pull request by chibenwa**
Here I want to answer the why questions : 
- Why not relying on Cassandra codec ? => https://github.com/datastax/java-driver/pull/656 => See below for userType
- Why userType ? Using user types do not clash our Cassandra initialisation model. Tuple does ! (We need to create Tuple before defining the tables that uses it. What is not required with UDT)

Then the implementation is pretty staightforward.

### CODE REVIEW

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/CodecProvider.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.backends.cassandra.init;
+
+import org.apache.james.backends.cassandra.utils.ZonedDateTimeCodec;
+
+import com.datastax.driver.core.Session;
+
+public class CodecProvider {
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/init/CodecProvider.java:26]: I don't understand this class

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+import com.datastax.driver.core.DataType;
+import com.datastax.driver.core.ParseUtils;
+import com.datastax.driver.core.ProtocolVersion;
+import com.datastax.driver.core.TypeCodec;
+import com.datastax.driver.core.UserType;
+import com.datastax.driver.core.exceptions.InvalidTypeException;
+
+public class ZonedDateTimeCodec extends TypeCodec.AbstractUDTCodec<ZonedDateTime> {
+
+    private static final java.time.format.DateTimeFormatter FORMATTER = new java.time.format.DateTimeFormatterBuilder()
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:45]: some pretty indentation would help

Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:45]: There's no standard formatter for the format you use ?

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    @Override
+    protected ZonedDateTime newInstance() {
+        return null;
+    }
+
+    @Override
+    protected ByteBuffer serializeField(ZonedDateTime zonedDateTime, String fieldName, ProtocolVersion protocolVersion) {
+        switch (fieldName) {
+            case CassandraZonedDateTimeModule.DATE:
+                return timestamp().serialize(new Date(zonedDateTime.toInstant().toEpochMilli()), protocolVersion);
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:89]: `Date.from(zonedDateTime.toInstant())`

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 2 above)
Comment (chibenwa) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:45]: This is only use for the format methods... It is taken from Datastax ZonedDateTimeCodec. I am sure they had great discussions about this....

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+            default:
+                throw new InvalidTypeException(fieldName + " is not known as a UDT field used for ZonedDateTime");
+        }
+    }
+
+    @Override
+    protected ZonedDateTime deserializeAndSetField(ByteBuffer byteBuffer, ZonedDateTime zonedDateTime, String fieldName, ProtocolVersion protocolVersion) {
+        switch (fieldName) {
+            case CassandraZonedDateTimeModule.DATE:
+                return ZonedDateTime.ofInstant(
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:101]: what if you don't get Date for the first call ?

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+        cassandra.getConf().execute(insertInto(DateRowModule.TABLE_NAME)
+            .value(DateRowModule.KEY, 18)
+            .value(DateRowModule.DATE, zonedDateTime));
+        ZonedDateTime retrievedZonedDateTime = cassandra.getConf()
+            .execute(select()
+                .from(DateRowModule.TABLE_NAME)
+                .where(eq(DateRowModule.KEY, 18)))
+            .one()
+            .get(DateRowModule.DATE, zonedDateTimeCodec);
+        if (zonedDateTime == null) {
```
Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java:101]: no if in tests

File: server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraSessionModule.java [hunk 6]
```diff
@@ -85,6 +87,17 @@ Cluster provideCluster(FileSystem fileSystem, AsyncRetryExecutor executor) throw
                 .get();
     }
 
+    @Provides
+    @Singleton
+    ZonedDateTimeCodec provideZonedDateTimeCodec(Session session, CassandraTypesProvider cassandraTypesProvider) {
```
Comment (mbaechler) [inline server/container/guice/cassandra-guice/src/main/java/org/apache/james/modules/mailbox/CassandraSessionModule.java:92]: It does much more than providing the codec, it also registers it. You should do only one thing at a time.

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+import org.apache.james.backends.cassandra.init.CassandraZonedDateTimeModule;
+
+import com.datastax.driver.core.DataType;
+import com.datastax.driver.core.ParseUtils;
+import com.datastax.driver.core.ProtocolVersion;
+import com.datastax.driver.core.TypeCodec;
+import com.datastax.driver.core.UserType;
+import com.datastax.driver.core.exceptions.InvalidTypeException;
+
+public class ZonedDateTimeCodec extends TypeCodec.AbstractUDTCodec<ZonedDateTime> {
```
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:43]: what is the test coverage of this class ?

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 2 above)
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:45]: Taken from another project with a CLA ? I heard nothing, but remove all of imported code fast !

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 4 above)
Comment (chibenwa) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:101]: I will. "protected" method so called from the super class. I have guaranties on the UDT type about field ordering and the caller iterate over fields in order. Want me to add a Preconditions to ensure Date is called first ?

(Apparently even datastax has API calls that looks uggly compared to ours...) 

https://github.com/datastax/java-driver/blob/3.0/driver-extras/src/main/java/com/datastax/driver/extras/codecs/jdk8/ZonedDateTimeCodec.java # deserializeAndSetField

```
        if (index == 0) {
            long millis = bigint().deserializeNoBoxing(input, protocolVersion);
            return Instant.ofEpochMilli(millis).atZone(ZoneOffset.UTC);
        }
        if (index == 1) {
            String zoneId = varchar().deserialize(input, protocolVersion);
            return target.withZoneSameInstant(ZoneId.of(zoneId));
        }
        throw new IndexOutOfBoundsException("Tuple index out of bounds. " + index);
```

Ouch ;-)

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java (hunk 5 above)
Comment (chibenwa) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java:101]: So can I PR assertj to make this possible ? 

```
assertThat(null).isEqualTo(null)
```

Otherwise, it ruins parameter use...

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 4 above)
Comment (mbaechler) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:101]: I think `newInstance()` is here to create a new instance before calling deserializeAndSet. That way, you don't need a precondition.

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 7 above)
Comment (chibenwa) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:43]: 50%... Ok.

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java (hunk 5 above)
Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java:101]: you can write other tests to handle edge case testing

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 3 above)
Comment (adutra) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:89]: Use `bigint().serializeNoBoxing(zonedDateTime.toInstant().toEpochMilli(), protocolVersion)` instead to avoid creating a temporary `Date` object.

File: backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java (hunk 4 above)
Comment (adutra) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:101]: > Apparently even datastax has API calls that looks uggly compared to ours...

Feel free to propose anything less ugly.

Comment (chibenwa) [inline backends-common/cassandra/src/main/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodec.java:101]: Sure, I will have a look into it.

File: backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java (hunk 5 above)
Comment (chibenwa) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java:101]: > you can write other tests to handle edge case testing

These edge cases will then get tested for each possible parameter, augmenting significatly the number of tests.

Comment (mbaechler) [inline backends-common/cassandra/src/test/java/org/apache/james/backends/cassandra/utils/ZonedDateTimeCodecTest.java:101]: If the tool is not the good one, change the tool, not the way you are supposed to write a test.
